[FLINK-31123][connector/files] Add IT cases for FileSink speculative …

Rruop opened the pull request:
## What is the purpose of the change

This change added an it case for FileSink in speculative execution mode that the sink will be scheduled by the speculative scheduler when there is a slow task.

## Brief change log

  - Added an test to assert that the FileSink will execute multiple attempts and the final results are correct when there is a slow sink and enable speculative scheduling.

## Verifying this change

This change is an it case.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

zhuzhurk commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/sink/FileSinkSpeculativeITCase.java:
```diff
@@ -0,0 +1,228 @@
+/**
+ * Tests {@link FileSink} with speculative scheduling enabled.
+ */
+public class FileSinkSpeculativeITCase {
```
```suggestion
class FileSinkSpeculativeITCase {
```

Rruop replied:
ok, I will correct it

zhuzhurk commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/sink/FileSinkSpeculativeITCase.java:
```diff
@@ -0,0 +1,228 @@
+    private static final AtomicInteger slowTaskCounter = new AtomicInteger(1);
+
+    @Before
+    public void setUp() {
```
```suggestion
void setUp() {
```

These methods are no longer required to be `public` in JUnit5.

Rruop replied:
ok, I will correct it

zhuzhurk commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/sink/FileSinkSpeculativeITCase.java:
```diff
@@ -0,0 +1,228 @@
+            if (jobVertex.getName().contains("test_map")) {
+                assertThat(jobVertex.getName().contains("file_sink")).isTrue();
+            }
+        }
```
It's better to move this check into `createJobGraph(...)`.

zhuzhurk commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/sink/FileSinkSpeculativeITCase.java:
```diff
@@ -0,0 +1,228 @@
+                new MiniClusterConfiguration.Builder()
+                        .withRandomPorts()
+                        .setNumTaskManagers(1)
+                        .setNumSlotsPerTaskManager(12)
```
I guess 4 slots are enough?

Rruop replied:
yes, 4 slot are enough for the job and I will correct that

zhuzhurk reviewed (commented):
Thanks for creating this PR! @Rruop 
I have a few minor comments. Please take a look.

zhuzhurk commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/sink/FileSinkSpeculativeITCase.java:
```diff
@@ -0,0 +1,228 @@
+
+    private static final AtomicInteger slowTaskCounter = new AtomicInteger(1);
+
+    @Before
```
`BeforeEach` of JUnit5 should be used.

Rruop replied:
done

Rruop commented:
> * @flinkbot run azure

Rruop commented:
@flinkbot run azure

zhuzhurk commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/sink/FileSinkSpeculativeITCase.java:
```diff
@@ -100,14 +96,6 @@ void testFileSinkSpeculative()
             miniCluster.executeJobBlocking(jobGraph);
```
It's better to use `miniClusterExtension` and execute the job direct with `StreamExecutionEnvironment#execute(...)`, instead of manually maintaining and using a `MiniCluster`. A reference can be `RateLimitedSourceReaderITCase`.

zhuzhurk replied:
The current way is also acceptable to me, except that the `@ExtendWith(MiniClusterExtension.class)` should be removed if we choose this way.

Rruop replied:
> The current way is also acceptable to me, except that the `@ExtendWith(MiniClusterExtension.class)` should be removed if we choose this way.
If  we remove @ExtendWith(MiniClusterExtension.class) and manually using a MiniCluster,  the ITCaseRules::ITCASE_USE_MINICLUSTER rule will be violated. Do you have other better way?

zhuzhurk replied:
Would it work by declaring the MiniClusterExtension as below?
```
    @RegisterExtension
    private static final MiniClusterExtension MINI_CLUSTER =
            new MiniClusterExtension(
                    new MiniClusterResourceConfiguration.Builder()
                            .setConfiguration(configure())
                            .setNumberTaskManagers(1)
                            .setNumberSlotsPerTaskManager(4)
                            .build());
```

Rruop replied:
Sorry, I just mixed up the configuration of the cluster and the job. The cluster is started according to the MiniClusterResourceConfiguration. It will work according to your suggestion. Thank you.

zhuzhurk reviewed (approved):
LGTM. Thanks for addressing the comments! @Rruop 
Would you squash the commits?

Rruop commented:
> LGTM. Thanks for addressing the comments! @Rruop Would you squash the commits?

Thanks for your patiently reviewing, I will squash all the commits into one.
